### PR TITLE
add functions for creating ray with oauth proxy in front of the dashboard

### PR DESCRIPTION
**Pull request by KPostOffice**
# Issue link
closes: #174 

# What changes have been made
I added functions which create the necessary objects and update the appwrapper so that the SDK can create a ray cluster with an OAuth Proxy in front of the dashboard.

- OAuth sidecar container and tls secret mount
- Ingress with annotations for tls passthrough
- Service which points to OAuth port in the sidecar container and annotations for autogenerated tls certs
- Service account which contains the OAuth Callback
- ClusterRoleBinding giving OAuth sidecar necessary permissions

# Verification steps
go through notebook but set `openshift_oauth` to true in the config

## Checks
- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] Testing is not required for this change

### CODE REVIEW

**Comment by KPostOffice** [comment]
This isn't quite finished yet. I'm having tls related issues between the Ray API and the OAuth proxy.

**Comment by KPostOffice** [comment]
Blocked due to required changes upstream with the Ray Job Submission Client and TorchX

**Comment by blublinsky** [comment]
This URL is used not only for dashboard access but also for the APIs - currently jobs and In future also  Serve. Oath proxy requires manual log in, which will break these APIs

**Comment by KPostOffice** [comment]
@blublinsky I've launched OAuth so that it supports both GUI manual login as well as authentication using an auth bearer token. Can you PTAL?

**Comment by blublinsky** [comment]
> @blublinsky I've launched OAuth so that it supports both GUI manual login as well as authentication using an auth bearer token. Can you PTAL?

How do you get the auth bearer token to use?

**Comment by KPostOffice** [comment]
> > @blublinsky I've launched OAuth so that it supports both GUI manual login as well as authentication using an auth bearer token. Can you PTAL?
> 
> How do you get the auth bearer token to use?

https://github.com/project-codeflare/codeflare-sdk/pull/298/files#diff-5706a96c5346dd54a22370211260411363c4ed41d798fddf240eb8983b1366eaR82-R86

It is the same token used for interacting with the kubernetes api with the kubectl client. I grab it here based on the currently logged in user or in cluster config. I believe it is the same token as the one you generate from the `copy login command` page in the OpenShift client.

**Comment by blublinsky** [comment]
> It is the same token used for interacting with the kubernetes api with the kubectl client. I grab it here based on the currently logged in user or in cluster config. I believe it is the same token as the one you generate from the `copy login command` page in the OpenShift client.

So basically the only thing that you are validating is that user is logged in. Not very secure, really

**Comment by KPostOffice** [comment]
> > It is the same token used for interacting with the kubernetes api with the kubectl client. I grab it here based on the currently logged in user or in cluster config. I believe it is the same token as the one you generate from the `copy login command` page in the OpenShift client.
> 
> So basically the only thing that you are validating is that user is logged in. Not very secure, really

https://github.com/project-codeflare/codeflare-sdk/pull/298/files#diff-279fd895cfcb8295f1b610826f8eb7a40fd543ff873661acd9dc452cef3d9d14R423

The OAuth Proxy supports checking for authorization based on RBAC. Here I check to make sure the authenticated user has authorization to get pods in the given namespace.

**Comment by blublinsky** [comment]
So it is basically a single cluster solution

**Comment by KPostOffice** [comment]
> So it is basically a single cluster solution

I'm unsure what you are expecting. How would you propose we authenticate for a multi-cluster design here? AFAIK, the `codeflare-sdk` expects the user to be logged into a specific k8 cluster.

**Comment by blublinsky** [comment]
> I'm unsure what you are expecting. How would you propose we authenticate for a multi-cluster design here? AFAIK, the `codeflare-sdk` expects the user to be logged into a specific k8 cluster.

I do not think authentication/authorization should be linked to a given cluster, It should rather be for a given user, who can be logged in any of the several clusters, but that's just my opinion

**Comment by KPostOffice** [comment]
I was running into issues with the generate CA cert from the service annotation. Putting this here so I remember to look into the ca-injector.

https://cert-manager.io/docs/concepts/ca-injector/

**Comment by KPostOffice** [comment]
> > I'm unsure what you are expecting. How would you propose we authenticate for a multi-cluster design here? AFAIK, the `codeflare-sdk` expects the user to be logged into a specific k8 cluster.
> 
> I do not think authentication/authorization should be linked to a given cluster, It should rather be for a given user, who can be logged in any of the several clusters, but that's just my opinion

Where do you propose authentication and authorization occur in this case?

File: src/codeflare_sdk/cluster/cluster.py [hunk 1]
```diff
@@ -61,6 +69,26 @@ def __init__(self, config: ClusterConfiguration):
         self.config = config
         self.app_wrapper_yaml = self.create_app_wrapper()
         self.app_wrapper_name = self.app_wrapper_yaml.split(".")[0]
+        self._client = None
+        self._k8_client = config.new_client_from_config()
```
Comment (MichaelClifford) [inline src/codeflare_sdk/cluster/cluster.py:73]: I think config is getting overloaded here. I just got the error: `AttributeError: 'ClusterConfiguration' object has no attribute 'new_client_from_config'`. 

It seems like you want to use `kubernetes.config` but just above you are setting `self.config` to the ClusterConfiguration object.

**Review by MichaelClifford — DISMISSED** [review]

File: src/codeflare_sdk/cluster/cluster.py (hunk 1 above)
Comment (KPostOffice) [reply src/codeflare_sdk/cluster/cluster.py:73]: Thanks, nice catch

**Comment by MichaelClifford** [comment]
Tested out starting a ray cluster from local machine and submitting jobs. Seems to work as expected when logging in via sdk authentication or `oc login`. 

lgtm

@Maxusmusti could you please review as well.

**Comment by anishasthana** [comment]
@KPostOffice could you create a follow-on task to switch this over to istio (once we have istio available ootb in ODH). Or at least make it configurable somehow. Alternatively, the follow-on winds up being "Remove this, and update the kuberay operator to allow auth enablement"

File: src/codeflare_sdk/utils/kube_api_helpers.py [hunk 2]
```diff
@@ -42,3 +43,7 @@ def _kube_api_error_handling(e: Exception):  # pragma: no cover
         elif e.reason == "Conflict":
             raise FileExistsError(exists_msg)
     raise e
+
+
+def _get_api_host(api_client: client.ApiClient):  # pragma: no cover
```
Comment (Bobbins228) [inline src/codeflare_sdk/utils/kube_api_helpers.py:48]: When creating the appwrapper while authenticated with a K8s config file I experienced an error `'NoneType' object has no attribute 'configuration'`

When I ran through these changes while authenticated with token and server and it ran as expected. Just need to fix the _get_api_host function

**Review by Bobbins228 — CHANGES_REQUESTED** [review]

**Review by Bobbins228 — APPROVED** [review]
Verified working on OSD cluster awesome work Kevin /approve

**Pull request merged**